# Práctica Clase 1 

1) Crear una función capaz de convertir números enteros de base 10 a base 2. Debe recibir como parámetro el número a convertir<br>
Consideraciones:<br> 
a. Tratar de resolverlo sin usar la función format(nro,"b")<br>
b. El pdf "conversion-de-decimal-a-binario.pdf" puede servir de ayuda.


In [1]:
def NumeroBinario(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    if (type(numero) != int):
        return None
    if (numero < 0):
        return None
    elif (numero == 0):
        return 0

    lista_binario = []
    div_resto = 0

    while(numero > 0):
        div_resto = numero % 2
        numero = numero // 2
        lista_binario.append(div_resto)

    numero_binario = ''
    for e in lista_binario:
        numero_binario = str(e) + numero_binario  

    return int(numero_binario)


Versión 2

In [17]:
def NumeroBinario(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    if type(numero) != int or numero < 0: #Con o sin paréntesis?
        return None
    elif numero == 0:
        return 0
    lista_binario = [] #Podríamos usar una tupla?
    div_resto = 0

    while numero > 0:
        div_resto = numero % 2
        numero = numero // 2
        lista_binario.append(div_resto) #Podría usar el método append si fuera una cadena de texto/tupla?

    numero_binario = ''
    for e in lista_binario:
        numero_binario = str(e) + numero_binario #Empieza desde el final

    return int(numero_binario)

In [18]:
num = 33
NumeroBinario(num), format(num, 'b')

(100001, '100001')

In [12]:
NumeroBinario(12), format(12, 'b')

(11, '1100')

In [13]:
#Ejemplo:
NumeroBinario(29) 

10111

2) Convertir de decimal a binario las fracciones 1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9. Luego analizar los resultados y observar qué particularidad se encuentra en los mismos. Se puede usar Python o una calculadora, lo importante es ver si hay algo que podemos notar...

In [56]:
def NumeroBinarioFraccionario(numero_decimal):
    if type(numero_decimal) != float or numero_decimal < 0 or numero_decimal > 1:
        return None
    elif numero_decimal == 0:
        return 0
    
    lim_periodico = 24
    lista_binario = []
    i = 0
    while numero_decimal > 0 and i < lim_periodico:
        lista_binario.append(int(numero_decimal * 2))
        numero_decimal = numero_decimal * 2 - int(numero_decimal * 2)
        i+=1

    numero_binario = '0.'
    for e in lista_binario:
        numero_binario += str(e)

    return(numero_binario)    

In [65]:
    numero_decimal = 1/3
    lim_periodico = 24
    lista_binario = []
    while numero_decimal > 0 and i < lim_periodico:
        lista_binario.append(int(numero_decimal * 2))
        numero_decimal = numero_decimal * 2 - int(numero_decimal * 2)
        print(i, lista_binario)
        i+=1

13 [0]
14 [0, 1]
15 [0, 1, 0]
16 [0, 1, 0, 1]
17 [0, 1, 0, 1, 0]
18 [0, 1, 0, 1, 0, 1]
19 [0, 1, 0, 1, 0, 1, 0]
20 [0, 1, 0, 1, 0, 1, 0, 1]
21 [0, 1, 0, 1, 0, 1, 0, 1, 0]
22 [0, 1, 0, 1, 0, 1, 0, 1, 0, 1]
23 [0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0]


In [62]:
for i in range(2,10):
    print(f"fraccion 1/{i}: {str(1/i)} y en binario: {NumeroBinarioFraccionario(1/i)}") #f-string
    #qué pasa con 1/5?



fraccion 1/2: 0.5 y en binario: 0.1
fraccion 1/3: 0.3333333333333333 y en binario: 0.010101010101010101010101
fraccion 1/4: 0.25 y en binario: 0.01
fraccion 1/5: 0.2 y en binario: 0.001100110011001100110011
fraccion 1/6: 0.16666666666666666 y en binario: 0.001010101010101010101010
fraccion 1/7: 0.14285714285714285 y en binario: 0.001001001001001001001001
fraccion 1/8: 0.125 y en binario: 0.001
fraccion 1/9: 0.1111111111111111 y en binario: 0.000111000111000111000111


In [79]:
for i in range(2,10):
    print('Fraccion 1 /', i, ': ', str(1/i), ' y En binario: ', NumeroBinarioFraccionario(1/i)) #puedo utilizar las llamadas f-strings?
    #qué pasa con 1/5?

Fraccion 1 / 2 :  0.5  y En binario:  0.1
Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
Fraccion 1 / 4 :  0.25  y En binario:  0.01
Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
Fraccion 1 / 8 :  0.125  y En binario:  0.001
Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111


In [83]:
NumeroBinarioFraccionario(0.1)

'0.000110011001100110011001'

Salida esperada:
* Fraccion 1 / 2 :  0.5  y En binario:  0.1
* Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
* Fraccion 1 / 4 :  0.25  y En binario:  0.01
* Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
* Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
* Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
* Fraccion 1 / 8 :  0.125  y En binario:  0.001
* Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111